File: cpp/LeetCode/No37.cpp
/*
Write a program to solve a Sudoku puzzle by filling the empty cells.

A sudoku solution must satisfy all of the following rules:

Each of the digits 1-9 must occur exactly once in each row.
Each of the digits 1-9 must occur exactly once in each column.
Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
The '.' character indicates empty cells.
*/
#include <iostream>
#include <vector>
#include "Debug\VectorDebug.h"

using namespace std;

class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        vector<vector<bool>> row = vector<vector<bool>>(9, vector<bool>(9));
        vector<vector<bool>> col = vector<vector<bool>>(9, vector<bool>(9));
        vector<vector<bool>> block = vector<vector<bool>>(9, vector<bool>(9));
        for(int r=0;r<9;r++){
            for(int c=0;c<9;c++){
                if(board[r][c]!='.'){
                    int index = board[r][c] - '1';
                    row[r][index] = true;
                    col[c][index] = true;
                    block[r/3*3+c/3][index] = true;
                }
            }
        }
        backtrack(board, row, col, block);
    }
private:
    bool backtrack(vector<vector<char>>& board, vector<vector<bool>> &row, vector<vector<bool>> &col, vector<vector<bool>> &block, int r=0, int c=0){
        bool result = false;
        if(c==9){
            r++;
            if(r==9){
                return true;
            }
            c=0;
        }
        
        if(board[r][c]=='.'){
            for(int n=0;n<9;n++){
                if(!row[r][n] && !col[c][n] && !block[r/3*3+c/3][n]){
                    board[r][c] = '1' + n;
                    row[r][n] = true;
                    col[c][n] = true;
                    block[r/3*3+c/3][n] = true;
                    refresh2DVector<char>(board);
                    result = backtrack(board, row, col, block, r, c+1);
                    if(!result){
                        board[r][c] = '.';
                        row[r][n] = false;
                        col[c][n] = false;
                        block[r/3*3+c/3][n] = false;
                    }
                    else{
                        return result;
                    }
                }
            }
        }
        else{
            result = backtrack(board, row, col, block, r, c+1);
        }
        return result;
    }
};

int main(void){
    Solution solution;

    vector<vector<char>> board{
                                {'5','3','.','.','7','.','.','.','.'},
                                {'6','.','.','1','9','5','.','.','.'},
                                {'.','9','8','.','.','.','.','6','.'},
                                {'8','.','.','.','6','.','.','.','3'},
                                {'4','.','.','8','.','3','.','.','1'},
                                {'7','.','.','.','2','.','.','.','6'},
                                {'.','6','.','.','.','.','2','8','.'},
                                {'.','.','.','4','1','9','.','.','5'},
                                {'.','.','.','.','8','.','.','7','9'}
                              };
    solution.solveSudoku(board);
    refresh2DVector<char>(board);
    return 0;
}

File: python/LeetCode Quiz/No37.py
'''
Write a program to solve a Sudoku puzzle by filling the empty cells.

A sudoku solution must satisfy all of the following rules:

Each of the digits 1-9 must occur exactly once in each row.
Each of the digits 1-9 must occur exactly once in each column.
Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
The '.' character indicates empty cells.
'''
from typing import List

def printBoard(board: List[List[str]]):
    for row in board:
        print(row)

class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        def backtrack(r=0, c=0):
            if c==9:
                r+=1
                if r==9:
                    self.__done = True
                    return
                c=0
            
            if board[r][c]!='.':
                backtrack(r, c+1)
            else:
                for n in range(9):
                    if self.__row[r][n] or self.__col[c][n] or self.__blo[r//3*3+c//3][n]:
                        continue
                    board[r][c] = chr(n+ord('1'))
                    self.__row[r][n] = True
                    self.__col[c][n] = True
                    self.__blo[r//3*3+c//3][n] = True
                    backtrack(r, c+1)
                    if not self.__done:
                        board[r][c] = '.'
                        self.__row[r][n] = False
                        self.__col[c][n] = False
                        self.__blo[r//3*3+c//3][n] = False
            
        self.__done = False
        self.__row = [[False]*9 for _ in range(9)]
        self.__col = [[False]*9 for _ in range(9)]
        self.__blo = [[False]*9 for _ in range(9)]
        for r in range(9):
            for c in range(9):
                if board[r][c]!='.':
                    index = int(board[r][c]) - int('1')
                    self.__row[r][index] = True
                    self.__col[c][index] = True
                    self.__blo[r//3*3+c//3][index] = True
        backtrack()

def main():
    solution = Solution()

    board = [["5","3",".",".","7",".",".",".","."],
             ["6",".",".","1","9","5",".",".","."],
             [".","9","8",".",".",".",".","6","."],
             ["8",".",".",".","6",".",".",".","3"],
             ["4",".",".","8",".","3",".",".","1"],
             ["7",".",".",".","2",".",".",".","6"],
             [".","6",".",".",".",".","2","8","."],
             [".",".",".","4","1","9",".",".","5"],
             [".",".",".",".","8",".",".","7","9"]]
    
    solution.solveSudoku(board)
    printBoard(board)

if __name__=='__main__':
    main()
